wallet/wallet.go

```diff
@@ -0,0 +1,522 @@
+	if amount.IsZero() {
+		return nil, func() {}, nil
+	}
+
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
+	utxos, err := sw.store.UnspentSiacoinElements()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tpoolSpent := make(map[types.Hash256]bool)
+	tpoolUtxos := make(map[types.Hash256]types.SiacoinElement)
+	for _, txn := range sw.cm.PoolTransactions() {
+		for _, sci := range txn.SiacoinInputs {
+			tpoolSpent[types.Hash256(sci.ParentID)] = true
+			delete(tpoolUtxos, types.Hash256(sci.ParentID))
+		}
+		for i, sco := range txn.SiacoinOutputs {
+			tpoolUtxos[types.Hash256(txn.SiacoinOutputID(i))] = types.SiacoinElement{
+				StateElement: types.StateElement{
+					ID: types.Hash256(types.SiacoinOutputID(txn.SiacoinOutputID(i))),
+				},
+				SiacoinOutput: sco,
+			}
+		}
+	}
+
+	// remove locked and spent outputs
+	usableUTXOs := utxos[:0]
+	for _, sce := range utxos {
+		if sw.locked[sce.ID] || tpoolSpent[sce.ID] {
+			continue
+		}
+		usableUTXOs = append(usableUTXOs, sce)
+	}
+
+	// sort by value, descending
+	sort.Slice(usableUTXOs, func(i, j int) bool {
+		return usableUTXOs[i].SiacoinOutput.Value.Cmp(usableUTXOs[j].SiacoinOutput.Value) > 0
+	})
+
+	// fund the transaction using the largest utxos first
+	var selected []types.SiacoinElement
+	var inputSum types.Currency
+	for i, sce := range usableUTXOs {
+		if inputSum.Cmp(amount) >= 0 {
+			usableUTXOs = usableUTXOs[i:]
+			break
+		}
+		selected = append(selected, sce)
+		inputSum = inputSum.Add(sce.SiacoinOutput.Value)
+	}
+
+	// if the transaction can't be funded, return an error
+	if inputSum.Cmp(amount) < 0 {
+		return nil, nil, ErrNotEnoughFunds
+	}
+
+	// check if remaining utxos should be defragged
+	txnInputs := len(txn.SiacoinInputs) + len(selected)
+	if len(usableUTXOs) > transactionDefragThreshold && txnInputs < maxInputsForDefrag {
+		// add the smallest utxos to the transaction
+		defraggable := usableUTXOs
+		if len(defraggable) > maxDefragUTXOs {
+			defraggable = defraggable[len(defraggable)-maxDefragUTXOs:]
+		}
+		for i := len(defraggable) - 1; i >= 0; i-- {
+			if txnInputs >= maxInputsForDefrag {
+				break
+			}
+
+			sce := defraggable[i]
+			selected = append(selected, sce)
+			inputSum = inputSum.Add(sce.SiacoinOutput.Value)
+			txnInputs++
+		}
+	}
+
+	// add a change output if necessary
+	if inputSum.Cmp(amount) > 0 {
+		txn.SiacoinOutputs = append(txn.SiacoinOutputs, types.SiacoinOutput{
+			Value:   inputSum.Sub(amount),
+			Address: sw.addr,
+		})
+	}
+
+	toSign := make([]types.Hash256, len(selected))
+	for i, sce := range selected {
+		txn.SiacoinInputs = append(txn.SiacoinInputs, types.SiacoinInput{
+			ParentID:         types.SiacoinOutputID(sce.ID),
+			UnlockConditions: types.StandardUnlockConditions(sw.priv.PublicKey()),
+		})
+		toSign[i] = types.Hash256(sce.ID)
+		sw.locked[sce.ID] = true
+	}
+
+	release := func() {
+		sw.mu.Lock()
+		defer sw.mu.Unlock()
+		for _, id := range toSign {
+			delete(sw.locked, id)
+		}
+	}
+	return toSign, release, nil
+}
+
```

Comment:
would be nice to include our `Redistribute` method.

The original looks like this:
```go
// Redistribute returns a transaction that redistributes money in the wallet by
// selecting a minimal set of inputs to cover the creation of the requested
// outputs. It also returns a list of output IDs that need to be signed.
func (w *SingleAddressWallet) Redistribute(cs consensus.State, outputs int, amount, feePerByte types.Currency, pool []types.Transaction) ([]types.Transaction, []types.Hash256, error) {
	w.mu.Lock()
	defer w.mu.Unlock()

	// build map of inputs currently in the tx pool
	inPool := make(map[types.Hash256]bool)
	for _, ptxn := range pool {
		for _, in := range ptxn.SiacoinInputs {
			inPool[types.Hash256(in.ParentID)] = true
		}
	}

	// fetch unspent transaction outputs
	utxos, err := w.store.UnspentSiacoinElements(false)
	if err != nil {
		return nil, nil, err
	}

	// check whether a redistribution is necessary, adjust number of desired
	// outputs accordingly
	for _, sce := range utxos {
		inUse := w.isOutputUsed(sce.ID) || inPool[sce.ID]
		matured := cs.Index.Height >= sce.MaturityHeight
		sameValue := sce.Value.Equals(amount)
		if !inUse && matured && sameValue {
			outputs--
		}
	}
	if outputs <= 0 {
		return nil, nil, nil
	}

	// desc sort
	sort.Slice(utxos, func(i, j int) bool {
		return utxos[i].Value.Cmp(utxos[j].Value) > 0
	})

	// prepare all outputs
	var txns []types.Transaction
	var toSign []types.Hash256

	for outputs > 0 {
		var txn types.Transaction
		for i := 0; i < outputs && i < redistributeBatchSize; i++ {
			txn.SiacoinOutputs = append(txn.SiacoinOutputs, types.SiacoinOutput{
				Value:   amount,
				Address: w.Address(),
			})
		}
		outputs -= len(txn.SiacoinOutputs)

		// estimate the fees
		outputFees := feePerByte.Mul64(uint64(len(encoding.Marshal(txn.SiacoinOutputs))))
		feePerInput := feePerByte.Mul64(BytesPerInput)

		// collect outputs that cover the total amount
		var inputs []SiacoinElement
		want := amount.Mul64(uint64(len(txn.SiacoinOutputs)))
		var amtInUse, amtSameValue, amtNotMatured types.Currency
		for _, sce := range utxos {
			inUse := w.isOutputUsed(sce.ID) || inPool[sce.ID]
			matured := cs.Index.Height >= sce.MaturityHeight
			sameValue := sce.Value.Equals(amount)
			if inUse {
				amtInUse = amtInUse.Add(sce.Value)
				continue
			} else if sameValue {
				amtSameValue = amtSameValue.Add(sce.Value)
				continue
			} else if !matured {
				amtNotMatured = amtNotMatured.Add(sce.Value)
				continue
			}

			inputs = append(inputs, sce)
			fee := feePerInput.Mul64(uint64(len(inputs))).Add(outputFees)
			if SumOutputs(inputs).Cmp(want.Add(fee)) > 0 {
				break
			}
		}

		// not enough outputs found
		fee := feePerInput.Mul64(uint64(len(inputs))).Add(outputFees)
		if sumOut := SumOutputs(inputs); sumOut.Cmp(want.Add(fee)) < 0 {
			// in case of an error we need to free all inputs
			w.releaseInputs(txns...)
			return nil, nil, fmt.Errorf("%w: inputs %v < needed %v + txnFee %v (usable: %v, inUse: %v, sameValue: %v, notMatured: %v)",
				ErrInsufficientBalance, sumOut.String(), want.String(), fee.String(), sumOut.String(), amtInUse.String(), amtSameValue.String(), amtNotMatured.String())
		}

		// set the miner fee
		txn.MinerFees = []types.Currency{fee}

		// add the change output
		change := SumOutputs(inputs).Sub(want.Add(fee))
		if !change.IsZero() {
			txn.SiacoinOutputs = append(txn.SiacoinOutputs, types.SiacoinOutput{
				Value:   change,
				Address: w.addr,
			})
		}

		// add the inputs
		for _, sce := range inputs {
			txn.SiacoinInputs = append(txn.SiacoinInputs, types.SiacoinInput{
				ParentID:         types.SiacoinOutputID(sce.ID),
				UnlockConditions: StandardUnlockConditions(w.priv.PublicKey()),
			})
			toSign = append(toSign, sce.ID)
			w.lastUsed[sce.ID] = time.Now()
		}

		txns = append(txns, txn)
	}

	return txns, toSign, nil
}
```

Reply:
I know I've wanted this function before, but can you remind me of the benefit of this vs `FundTransaction` with ten siacoin outputs? Is it only the miner fee calculation?

Reply:
If you have 30 outputs of value X and you want 50, `Redistribute` can avoid touching the 30 outputs and create 20 more. 
`FundTransaction` with `n=20` might reuse the 30 outputs we already have so we need to do `n=50` which might result in anywhere from 50 to 70 outputs of the value we want.

Reply:
Going to handle this in a follow-up